### DISCUSSION TITLE
Changing mode from command control to hover assist mode.

### DISCUSSION THREAD

**Discussion by Residualstress** [bitcraze/discussions#1365] [Q&A] [answered]
Hi, I want to implement a function that change command control to hover assist mode and control it by the joystick manually for the safety.
For example, when the drone is doing a mission fly from A to B using below code:
```
static void setHoverSetpoint(setpoint_t *setpoint, float vx, float vy, float z)
{
  setpoint->mode.z = modeAbs;
  setpoint->position.z = z;
  
  setpoint->mode.x = modeVelocity;
  setpoint->mode.y = modeVelocity;
  setpoint->velocity.x = vx;
  setpoint->velocity.y = vy;
  
  setpoint->velocity_body = true;
}

setHoverSetpoint(&setpoint, velMax, 0, height_takeoff);
```
Is that possible to stop the command and change to hover assist mode immediately?
Now, I found when I controlled the drone by commands, the drone refused the control from the joystick.

**Comment by knmcguire** [comment]
We need some more information

So when you are trying to control the drone with the joystick, is the function or app still constantly sending this setpoint? Is this changed directly in the firmware or as an app layer?

**Reply by Residualstress** [reply]
Hi, Knmcguire! Thanks for your reply.

Actually, I can start and stop the app layer manually rather than execute at boot time automatically. So I guess it would be easier to changed directly in the firmware.

Once I stop the app layer, the setpoint from it will not be sent immediately. But I could code in [app_handler.c](https://github.com/bitcraze/crazyflie-firmware/blob/a25e8fb2254f6d5be54ad57566b3b846a97de93b/src/modules/src/app_handler.c#L4) and continue sending setpoint to make the drone hover at the position where the application stops, for example:

```
logVarId_t idZ = logGetVarId("stateEstimate", "z");    
double position_z = logGetFloat(idZ);
setHoverSetpoint(&setpoint, 0, 0, position_z);
```
So, in such situation, can I take over the drone with a joystick and control it manually? Of course, the drone must stop recieving setpoint commands.

**Reply by knmcguire** [reply]
well technically as long as the setpoint of the takeover has a higher priority, you can take over the drone like that as well. Mind that the commanderSetSetpoint() is the function that officially pushes the setpoint to the commander framework: https://github.com/bitcraze/crazyflie-firmware/blob/a25e8fb2254f6d5be54ad57566b3b846a97de93b/src/modules/src/commander.c#L74

However, if you stop sending the take over setpoints, the lower priority will take over, so safest is to stop sending the autonomy setpoints as well.

**Reply by Residualstress** [reply]
> well technically as long as the setpoint of the takeover has a higher priority, you can take over the drone like that as well. Mind that the commanderSetSetpoint() is the function that officially pushes the setpoint to the commander framework: https://github.com/bitcraze/crazyflie-firmware/blob/a25e8fb2254f6d5be54ad57566b3b846a97de93b/src/modules/src/commander.c#L74
> 
> However, if you stop sending the take over setpoints, the lower priority will take over, so safest is to stop sending the autonomy setpoints as well.

Yes, when I press the button on the joystick, the autonomy setpoints in app layer will stop and the take over setpoints in app_handler.c will run immediately.  That's what I already could do now.

But I want to control it by the joystick after the autonomous mission in app layer. I have tried many times that once I sent setpoints to CF, my joystick couldn't control the drone anymore even if the sending of setpoints was over. 

Maybe my previous statement was ambiguous. I hope that when I stop the mission in app layer, the drone could changed to assist Hover mode, and then I can manually control the drone by the joystick.

**Reply by knmcguire** [reply, marked as answer]
Ah oke, then perhaps you'll need to enable to relax the priorities.

In the python api that is this function: `def send_notify_setpoint_stop(self, remain_valid_milliseconds=0)` (see https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/commander/)

That enables this line in the firmware: https://github.com/bitcraze/crazyflie-firmware/blob/a25e8fb2254f6d5be54ad57566b3b846a97de93b/src/modules/src/crtp_commander.c#L102

See if that works and let us know.

**Reply by Residualstress** [reply]
> Ah oke, then perhaps you'll need to enable to relax the priorities.
> 
> In the python api that is this function: `def send_notify_setpoint_stop(self, remain_valid_milliseconds=0)` (see https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/commander/)
> 
> That enables this line in the firmware: https://github.com/bitcraze/crazyflie-firmware/blob/a25e8fb2254f6d5be54ad57566b3b846a97de93b/src/modules/src/crtp_commander.c#L102
> 
> See if that works and let us know.

Thank you, knmcguire! The `commanderRelaxPriority()` did work. Since I used `app_handler.c` to start and stop app layer function, I added this command in this C file. Now I can use joystick to quit the  autonomous mission in app layer and take over the drone in hover assisted mode manually. This is really helpful to avoid crashes or stop abnormal behavior during the flight.

**Reply by knmcguire** [reply]
great! if you have any more problems you are hitting against, you can start a new discussion thread. Happy hacking!
